src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,171 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value, this.tail, null);
+
+        if (this.size == 0) {
+            this.head = newNode;
+        } else {
+            this.tail.next = newNode;
+        }
+
+        this.tail = newNode;
+        this.size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == this.size) {
+            add(value);
+            return;
+        }
+
+        Node<T> node = getNodeByIndex(index);
+        Node<T> newNode = new Node<>(value, node.prev, node);
+
+        if (node.prev != null) {
+            node.prev.next = newNode;
+        }
+
+        node.prev = newNode;
+
+        if (index == 0) {
+            head = newNode;
+        }
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldValue;
+
+        oldValue = node.value;
+        node.value = value;
+
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldValue;
+
+        oldValue = node.value;
+        updateHeadAndTail(node, index);
+        unlink(node);
+
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        int currentIndex = 0;
+        Node<T> currentNode = this.head;
+
+        if (this.size == 0) {
+            return false;
+        }
+
+        do {
+            if ((object == null)
+                    ? (currentNode.value == null)
+                    : (object.equals(currentNode.value))) {
+                updateHeadAndTail(currentNode, currentIndex);
```

Comment:
move this logic to unlink method, you can check if the node is head or tail by comparing reference ```node == head```

Reply:
+